### ISSUE TITLE
Control flow declarations in parenthesis are a syntax error

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#18435]
A common idiom in languages that have nillable objects and '?' and '!' operators (i.e. Swift) is to write an 'if' statement that does a nillability check and assigns a non-nillable version to a variable shadowing the nillable one.  As of 1.24 Chapel also supports this syntax (see the “Control-Flow Declarations” section of the release notes): https://chapel-lang.org/releaseNotes/1.24/01-lang-lib.pdf

This allows users to write code like this:

```chpl
class MyClass {
  const a = 10;
}

var x : owned MyClass? = new MyClass();
if var x = x  {
  writeln(x.a);
} else {
  writeln("X is nil!");
}
```

Funnily enough though, if you write the control flow declaration in parentheses you'll get a syntax error:

```chpl
if(var x = x) {
  writeln(x.a);
} else {
  writeln("X is nil!");
}
```

`foo.chpl:51: syntax error: near 'var'`

Users coming from a C background are likely to put parenthesis around the condition.

Arguably this is a bug. At the least I would call it a usability issue.

**Comment by mppf** [comment]
I'd expect it not to be hard to update the parser to handle it.

**Comment by bradcray** [comment]
My guess is that this was just an oversight in the implementation by @vasslitvinov rather than an intentional choice, but I'm tagging him to see.

Since the parenthesis in Chapel's conditional statements are not required, but just an instance of Chapel's normal expression parenthesization, would we want to support, say, `if (((((var x = y)))))` (which could suggest supporting `var` declarations in arbitrary expression productions even though I don't think it makes sense in most other contexts?), or simply `if (var x = y)` (which seems like it could be expressed as a simple addition to the conditional syntax productions).  I can't come up with a case where more complex parenthesizations would be required, but may just be lazy in my thinking.

Actually... would we want to support conditionals that have compound de-niling expressions like:

```chapel
if var xnn = x && var ynn = y then ...
```

or:

```chapel
if var xnn = x || ynn = y then ...
```

I suppose the latter doesn't really make sense because it isn't clear what the other variable would be within the context of the `then` clause.  And the former could be expressed through nested conditionals if needed.  What does Swift do here?  [you can treat that as rhetorical, I just haven't gone to look yet].

**Comment by stonea** [comment]
> Actually... would we want to support conditionals that have compound de-niling expressions like:

If we want to follow the precedent set by Swift it looks they're syntax supports it (copied from:
https://www.andrewcbancroft.com/2017/04/19/how-to-unwrap-multiple-optionals-with-one-if-let-in-swift/)

```swift
var greetings: [String]? = ["Howdy!", "Hello!"]
var salutations: [String]? = ["Hi!", "Hey!"]
if let g = greetings, let s = salutations {
    print(g)
    print(s)
}
```

**Comment by bradcray** [comment]
I see, so if I'm understanding, the comma essentially acts like a logical && for this situation (i.e., if either were nil, we wouldn't enter the conditional).  That might suggest doing something like this in Chapel:

```chapel
if const xnn = x, ynn = y then ...
```

to reflect our compound declaration syntax in normal situations:

```chapel
const c1 = expr1,
      c2 = expr2,
      c3 = expr3;
```

Rather than, say:

```chapel
if const xnn = x, const ynn = y then ...
```

A downside to this approach is that it wouldn't permit you to have a mix of non-nilable `const` and `var` declarations in a single conditional, say.  I don't know how limiting that would actually be in practice, though (and nested conditionals could be a fallback).

**Comment by vasslitvinov** [comment]
Indeed, handling only a paren-less form was not an intentional choice.

It makes sense to me to extend the syntax to allow a single level of parens (for people coming from C) and a compound declaration as Brad suggests: `if const xnn = x, ynn = y then ...` (to follow Swift's lead).
